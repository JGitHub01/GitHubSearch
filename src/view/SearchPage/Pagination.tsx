import { useState, useEffect } from 'react';

interface IPaginationProps {
  page?: number
  totalPages?: number
  numPagingButtons?: number
  onPaging?: (page: number) => void
}
export default function Pagination({ page = 1, totalPages = 1, numPagingButtons = 5, onPaging }: IPaginationProps) {
  const [pageRange, handleLeftPaging, handleRightPaging] = usePaging(page, totalPages, numPagingButtons, onPaging);
  const pageButtons = [];
  for (let k = pageRange[0]; k <= pageRange[1]; k++) {
    const pageButtonClasses = `pagination-page${k === page ? ' pagination-page--active' : ''}`;
    pageButtons.push(
      <button className={pageButtonClasses}
        onClick={() => { onPaging && onPaging(k); }} key={k}
      >
        {k}
      </button>)
  }

  return (
    totalPages === 1 ? null :
      <div className='pagination' data-testid='pagination'>
        <button className='pagination-left'
          disabled={pageRange[0] <= 1}
          onClick={handleLeftPaging}
          data-testid='pagination-left'
        >{'<'}</button>
        {pageButtons}
        <button className='pagination-right'
          disabled={pageRange[1] >= totalPages}
          onClick={handleRightPaging}
          data-testid='pagination-right'
        >{'>'}</button>
      </div>
  );
}

function usePaging(page: number, totalPages: number, numPagingButtons: number, onPaging?: (page: number) => void):
  [Array<number>, () => void, () => void] {
  const [pageRange, setPageRange] = useState<number[]>(() => {
    return calPageRange(page, totalPages, numPagingButtons);
  });

  const rangeLeft = pageRange[0];
  const rangeRight = pageRange[1];
  useEffect(() => {
    const newPagesInRange = Math.min(numPagingButtons, totalPages);
    if ((newPagesInRange !== rangeRight - rangeLeft + 1)
      || page < rangeLeft || page > rangeRight) {
      setPageRange(calPageRange(page, totalPages, numPagingButtons));
    }
  }, [page, rangeLeft, rangeRight, totalPages, numPagingButtons]);

  const handleLeftPaging = () => {
    if (pageRange[0] <= 1) {
      return;
    }
    const shift = Math.max(1, Math.floor(numPagingButtons / 2));
    const newPage = pageRange[0] - 1;
    const left = Math.max(1, newPage - shift);
    const right = Math.min(left + numPagingButtons - 1, totalPages);
    setPageRange([left, right]);
    onPaging && onPaging(newPage);
  }
  const handleRightPaging = () => {
    if (pageRange[1] >= totalPages) {
      return;
    }
    const shift = Math.max(1, Math.floor(numPagingButtons / 2));
    const newPage = pageRange[1] + 1;
    const right = Math.min(totalPages, newPage + shift);
    const left = Math.max(right - numPagingButtons + 1, 1);
    setPageRange([left, right]);
    onPaging && onPaging(newPage);
  }

  return [pageRange, handleLeftPaging, handleRightPaging];
}

function calPageRange(page: number, totalPages: number, numPagingButtons: number) {
  const pagesInRange = Math.min(numPagingButtons, totalPages);
  if (page >= 1 && page <= pagesInRange) {
    return [1, pagesInRange];
  }

  const right = Math.min(totalPages, page + Math.floor((numPagingButtons - 1) / 2));
  return [right - pagesInRange + 1, right];
}
