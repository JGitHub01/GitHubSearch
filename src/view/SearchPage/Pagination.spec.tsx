import {render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Pagination from './Pagination';

describe('Pagination', () => {
  it('Should render the basic snapshot view', () => {
    const {getByTestId} = render(<Pagination totalPages={5} />);
    const pagination = getByTestId('pagination');
    expect(pagination).toBeInTheDocument();

    expect(pagination).toMatchSnapshot();
  });

  it('Should render the snapshot view of 3 visible pages with 5 total pages', () => {
    const {getByTestId} = render(<Pagination page={4} totalPages={5} numPagingButtons={3} />);
    const pagination = getByTestId('pagination');
    expect(pagination).toBeInTheDocument();

    expect(pagination).toMatchSnapshot();
  });

  it('Should render the snapshot view at page 3', () => {
    const {getByTestId} = render(<Pagination totalPages={10} page={3} />);
    const pagination = getByTestId('pagination');
    expect(pagination).toBeInTheDocument();

    expect(pagination).toMatchSnapshot();
  });

  it('Should render the snapshot view at page 7', () => {
    const {getByTestId} = render(<Pagination totalPages={10} page={7} />);
    const pagination = getByTestId('pagination');
    expect(pagination).toBeInTheDocument();

    expect(pagination).toMatchSnapshot();
  });

  it('Should render the snapshot view at page 10', () => {
    const {getByTestId} = render(<Pagination totalPages={10} page={10} />);
    const pagination = getByTestId('pagination');
    expect(pagination).toBeInTheDocument();

    expect(pagination).toMatchSnapshot();
  });

  it('Should call onPaging callback with the new page number when a new page is clicked', () => {
    const callback = jest.fn();
    const {getByText} = render(<Pagination totalPages={5} onPaging={callback}/>);
    const page3 = getByText('3');
    userEvent.click(page3);
    expect(callback).toBeCalledWith(3);
  });

  it('Should call onPaging callback with the correct page number when the left paging button is clicked', () => {
    const callback = jest.fn();
    const {getByTestId} = render(<Pagination page={4} totalPages={7} numPagingButtons={3} onPaging={callback}/>);
    const leftButton = getByTestId('pagination-left');
    userEvent.click(leftButton);
    expect(callback).toBeCalledWith(2);
  });

  it('Should call onPaging callback with the correct page number when the right paging button is clicked', () => {
    const callback = jest.fn();
    const {getByTestId} = render(<Pagination totalPages={7} numPagingButtons={3} onPaging={callback}/>);
    const rightButton = getByTestId('pagination-right');
    userEvent.click(rightButton);
    expect(callback).toBeCalledWith(4);
  });
});
