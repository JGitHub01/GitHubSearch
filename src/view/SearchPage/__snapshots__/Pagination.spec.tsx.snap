// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render the basic snapshot view 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="pagination-left"
    data-testid="pagination-left"
    disabled=""
  >
    &lt;
  </button>
  <button
    class="pagination-page pagination-page--active"
  >
    1
  </button>
  <button
    class="pagination-page"
  >
    2
  </button>
  <button
    class="pagination-page"
  >
    3
  </button>
  <button
    class="pagination-page"
  >
    4
  </button>
  <button
    class="pagination-page"
  >
    5
  </button>
  <button
    class="pagination-right"
    data-testid="pagination-right"
    disabled=""
  >
    &gt;
  </button>
</div>
`;

exports[`Pagination Should render the snapshot view at page 3 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="pagination-left"
    data-testid="pagination-left"
    disabled=""
  >
    &lt;
  </button>
  <button
    class="pagination-page"
  >
    1
  </button>
  <button
    class="pagination-page"
  >
    2
  </button>
  <button
    class="pagination-page pagination-page--active"
  >
    3
  </button>
  <button
    class="pagination-page"
  >
    4
  </button>
  <button
    class="pagination-page"
  >
    5
  </button>
  <button
    class="pagination-right"
    data-testid="pagination-right"
  >
    &gt;
  </button>
</div>
`;

exports[`Pagination Should render the snapshot view at page 7 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="pagination-left"
    data-testid="pagination-left"
  >
    &lt;
  </button>
  <button
    class="pagination-page"
  >
    5
  </button>
  <button
    class="pagination-page"
  >
    6
  </button>
  <button
    class="pagination-page pagination-page--active"
  >
    7
  </button>
  <button
    class="pagination-page"
  >
    8
  </button>
  <button
    class="pagination-page"
  >
    9
  </button>
  <button
    class="pagination-right"
    data-testid="pagination-right"
  >
    &gt;
  </button>
</div>
`;

exports[`Pagination Should render the snapshot view at page 10 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="pagination-left"
    data-testid="pagination-left"
  >
    &lt;
  </button>
  <button
    class="pagination-page"
  >
    6
  </button>
  <button
    class="pagination-page"
  >
    7
  </button>
  <button
    class="pagination-page"
  >
    8
  </button>
  <button
    class="pagination-page"
  >
    9
  </button>
  <button
    class="pagination-page pagination-page--active"
  >
    10
  </button>
  <button
    class="pagination-right"
    data-testid="pagination-right"
    disabled=""
  >
    &gt;
  </button>
</div>
`;

exports[`Pagination Should render the snapshot view of 3 visible pages with 5 total pages 1`] = `
<div
  class="pagination"
  data-testid="pagination"
>
  <button
    class="pagination-left"
    data-testid="pagination-left"
  >
    &lt;
  </button>
  <button
    class="pagination-page"
  >
    3
  </button>
  <button
    class="pagination-page pagination-page--active"
  >
    4
  </button>
  <button
    class="pagination-page"
  >
    5
  </button>
  <button
    class="pagination-right"
    data-testid="pagination-right"
    disabled=""
  >
    &gt;
  </button>
</div>
`;
